### PR TITLE
Outsource mask calculation to new thread to decouple from clipping

### PR DESCRIPTION
**Pull request by peckto**
This PR outsources the mask calculation via tensorflow from the actual background clipping process.
In this way we can decouple the (slow) ANN task from the (fast) clipping task.
This results in a real time video rendering, even if the ANN task is slow(er).
As a side effect, the mask might be slower updated then the people moving the the video.
After all, this makes a smoother user experience, even with slower, but more stable models like `deeplabv3_257_mv_gpu`.

### CODE REVIEW

**Comment by floe** [comment]
Thanks, good suggestion - but that's a really big PR to review, do you think we can split it (e.g. first refactor the main chunk of code into a separate method, and then introduce the second thread)?

**Comment by floe** [comment]
(Now that I think about it, would it maybe make sense to not introduce a third thread, but rather have the mask calculation just run on the capture thread, and the NN calculation on the main thread?)

File: deepseg.cc [hunk 1]
```diff
@@ -403,21 +450,27 @@ int main(int argc, char* argv[]) {
...
-		fflush(stdout);
-		if (debug < 2) continue;
-
 		cv::Mat test;
 		cv::cvtColor(raw,test,CV_YUV2BGR_YUYV);
 		cv::imshow("output.png",test);
 		if (cv::waitKey(1) == 'q') break;
+
+		// throttle main trhead to save performance for tensorflow threads
+		usleep(100000);
```
Comment (BenBE) [inline deepseg.cc:459]: Maybe use condition variables or some event monitoring library like libev to handle the multiplexing …

File: deepseg.cc [hunk 2]
```diff
@@ -53,11 +53,11 @@ cv::Mat convert_rgb_to_yuyv( cv::Mat input ) {
 // Tensorflow Lite helper functions
 using namespace tflite;
 
-#define TFLITE_MINIMAL_CHECK(x)                              \
-  if (!(x)) {                                                \
+#define TFLITE_MINIMAL_CHECK(x)                               \
+	if (!(x)) {                                              \
 	fprintf(stderr, "Error at %s:%d\n", __FILE__, __LINE__); \
 	exit(1);                                                 \
```
Comment (BenBE) [inline deepseg.cc:59]: Shouldn't these gain one level of indentation?

File: deepseg.cc [hunk 3]
```diff
@@ -53,11 +53,11 @@ cv::Mat convert_rgb_to_yuyv( cv::Mat input ) {
 // Tensorflow Lite helper functions
 using namespace tflite;
 
-#define TFLITE_MINIMAL_CHECK(x)                              \
-  if (!(x)) {                                                \
+#define TFLITE_MINIMAL_CHECK(x)                               \
```
Comment (BenBE) [inline deepseg.cc:56]: Maybe be consistent with the alignment at the right?

**Comment by peckto** [comment]
Thank you for your feedback.
Sure, I will provide a second PR for splitting up the NN and rendering task is separate functions.

About your idea with the capture thread, would be possible. In this way we could get rid of the third thread.
I'm just not so sure, if it's a cleaner separation with the extra thread. 
Maybe we will see after refactoring the code.

**Comment by peckto** [comment]
And sorry for the code style issues, I will check my IDE settings.

**Comment by floe** [comment]
Continuing in #41 for now.

**Pull request closed without merging**